### PR TITLE
Feat(outcomes): Aggregate client reports [INGEST-247]

### PR DESCRIPTION
**Pull request by jjbayer**
Aggregate outcomes generated by client reports before sending them on.

The key for aggregation buckets consists of
* the outcome timestamp rounded to the bucket interval (1 minute default),
* all other outcome fields (`event_id` is `None` for client reports).

The value of an aggregated bucket is the outcome quantity.

This PR prepares for [INGEST-247](https://getsentry.atlassian.net/browse/INGEST-247). It does not yet implement collecting outcomes in external relays.

### CODE REVIEW

File: relay-server/src/actors/outcome_aggregator.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+                        scoping,
+                        outcome,
+                        event_id,
+                        remote_addr,
+                        category,
+                        quantity,
+                    };
+
+                    relay_log::trace!("Flushing outcome for timestamp {}", timestamp);
+                    outcome_producer.do_send(outcome).ok(); // TODO: should we handle send errors here?
```
Comment (jjbayer) [inline relay-server/src/actors/outcome_aggregator.rs]: Here we send messages in a loop. Could this overflow the outcome producer's mailbox? Should we send a batch of outcomes to the producer instead?

Comment (mitsuhiko) [reply relay-server/src/actors/outcome_aggregator.rs]: Jan told me that this bypasses the inbox

Comment (jjbayer) [reply relay-server/src/actors/outcome_aggregator.rs]: Right, thank you. The docs on `Recipient` are actually very clear on this:
https://docs.rs/actix/0.7.9/actix/struct.Recipient.html#method.do_send

File: relay-server/src/actors/outcome_aggregator.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+/// Inspired by [`relay_metrics::Aggregator`].
+pub struct OutcomeAggregator {
+    /// The width of each aggregated bucket in seconds
+    bucket_interval: u64,
+    /// The time we should wait before flushing a bucket, in seconds
+    flush_delay: u64,
+    /// Maximum capacity of the actor's inbox
+    mailbox_size: usize,
+    /// Mapping from offset to bucket key to quantity. timestamp = offset * bucket_interval
+    buckets: BTreeMap<u64, HashMap<BucketKey, u32>>,
```
Comment (untitaker) [inline relay-server/src/actors/outcome_aggregator.rs]: why not put the timestamp into the bucket key and avoid this nesting and extra allocation? Together with `split_off` optimization this should make the flush quite fast as long as the timestamp is declared on top of `BucketKey` (see documentation for `derive(PartialOrd)`)

File: relay-server/src/actors/outcome_aggregator.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            buckets: BTreeMap::new(),
+            outcome_producer,
+        }
+    }
+
+    fn do_flush(&mut self) {
+        let max_offset = (UnixTimestamp::now().as_secs() - self.flush_delay) / self.bucket_interval;
+        let bucket_interval = self.bucket_interval;
+        let outcome_producer = self.outcome_producer.clone();
+        self.buckets.retain(|offset, mapping| {
```
Comment (untitaker) [inline relay-server/src/actors/outcome_aggregator.rs]: you could probably use split_off here

**Review by untitaker — APPROVED** [review]
i think this does some unnecessary allocations on hashmap but otherwise it should be fine.

File: relay-config/src/config.rs [hunk 4]
```diff
@@ -946,6 +950,8 @@ impl Default for Outcomes {
             batch_size: 1000,
             batch_interval: 500,
             source: None,
+            bucket_interval: 60,
+            flush_delay: 30,
```
Comment (RaduW) [inline relay-config/src/config.rs]: Why do we flush every 30 seconds if the buckets are of 60 seconds ?

File: relay-server/src/actors/outcome_aggregator.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+            if offset <= &max_offset {
+                for (bucket_key, quantity) in mapping.drain() {
```
Comment (RaduW) [inline relay-server/src/actors/outcome_aggregator.rs]: I would not bother with this, I would just flush everything every `flush_delay`.
I understand that you don't want to send twice buckets that are not full yet but if we make the 
flush_delay >= bucket_internal (which I think we should) that would not be a big problem, we
may send buckets more than once anyway (for outcmoes that come late).

File: relay-server/src/actors/outcome_aggregator.rs (hunk 2 above)
Comment (RaduW) [reply relay-server/src/actors/outcome_aggregator.rs]: I agree with untitaker, I would use a single bucket_key -> quantity map (with the `timesamp % bucket_interval`  embedded in the bucket_key)

**Comment by RaduW** [comment]
Looks good.
I think we are overcomplicating a bit the logic of sending buckets but otherwise looks good.

**Review by RaduW — APPROVED** [review]

File: relay-server/src/actors/outcome_aggregator.rs (hunk 2 above)
Comment (jjbayer) [reply relay-server/src/actors/outcome_aggregator.rs]: Did as you suggested. Had to tweak the `BucketKey` type a bit in order to call `split_off` with a key containing only an offset.

File: relay-config/src/config.rs (hunk 4 above)
Comment (jjbayer) [reply relay-config/src/config.rs]: `flush_delay` is the grace period during which outcomes may still be submitted to a bucket after its time window has passed.

File: relay-server/src/actors/outcome_aggregator.rs (hunk 5 above)
Comment (jjbayer) [reply relay-server/src/actors/outcome_aggregator.rs]: I must admit I did not even think of this option, and I like the simplicity. Should have read this comment before implementing the `split_off` :)

Comment (jjbayer) [reply relay-server/src/actors/outcome_aggregator.rs]: @untitaker just to get an additional opinion, what do you think about the suggestion of simply flushing all the buckets on every flush?

Comment (untitaker) [reply relay-server/src/actors/outcome_aggregator.rs]: makes sense!

Comment (jjbayer) [reply relay-server/src/actors/outcome_aggregator.rs]: OK, simplified to use a single hashmap that gets flushed unconditionally every two minutes.

File: relay-server/src/actors/outcome_aggregator.rs [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+/// Aggregates outcomes into buckets, and flushes them periodically.
+/// Inspired by [`relay_metrics::Aggregator`].
+pub struct OutcomeAggregator {
+    /// The width of each aggregated bucket in seconds
+    bucket_interval: u64,
+    /// The number of seconds between flushes of all buckets
+    flush_interval: u64,
+    /// Mapping from bucket key to quantity.
+    buckets: HashMap<BucketKey, u32>,
```
Comment (untitaker) [inline relay-server/src/actors/outcome_aggregator.rs:43]: why do we use a hashmap here? (and also in the metrics aggregator I guess)

**Review by untitaker — APPROVED** [review]

File: relay-server/src/actors/outcome_aggregator.rs (hunk 6 above)
Comment (untitaker) [reply relay-server/src/actors/outcome_aggregator.rs:43]: recap from slack: probably no strong reason one way or the other right now

**Pull request merged**